src/Compiler/Driver/fsc.fs

```diff
@@ -162,17 +163,69 @@ let TypeCheck
 
         let eagerFormat (diag: PhasedDiagnostic) = diag.EagerlyFormatCore true
 
-        CheckClosedInputSet(
-            ctok,
-            diagnosticsLogger.CheckForErrors,
-            tcConfig,
-            tcImports,
-            tcGlobals,
-            None,
+        if false then
+            // I don't know yet if this is the right thing even
+            let assembly = (tcImports.DllTable.TryFind tcConfig.primaryAssembly.Name).Value
+
+            // these should be restored from raw resources probably
+            let byteReaderA () = ByteMemory.Empty.AsReadOnly()
+            let byteReaderB = None
+
+            let tcInfo = 
+                GetTypecheckingData(
+                    assembly.FileName,
+                    assembly.ILScopeRef,
+                    assembly.RawMetadata.TryGetILModuleDef(),
+                    byteReaderA,
+                    byteReaderB)
+
+            let rawData = tcInfo.RawData
+            let topAttrs = {
+                mainMethodAttrs = rawData.MainMethodAttrs
+                netModuleAttrs = rawData.NetModuleAttrs
+                assemblyAttrs = rawData.AssemblyAttrs
+            }
+
+            // need to understand where if anything can be used here, pickling state is hard
             tcInitialState,
-            eagerFormat,
-            inputs
-        )
+            topAttrs, 
+            rawData.DeclaredImpls,
+            // this is quite definitely wrong, need to figure out what to do with the environment
+            tcInitialState.TcEnvFromImpls
+
+        else
+            let tcState, topAttrs, declaredImpls, tcEnvAtEndOfLastFile = CheckClosedInputSet(
+                ctok,
+                diagnosticsLogger.CheckForErrors,
+                tcConfig,
+                tcImports,
+                tcGlobals,
+                None,
+                tcInitialState,
+                eagerFormat,
+                inputs
+            )
+
+            if false then
+                let tcInfo = {
+                    MainMethodAttrs = topAttrs.mainMethodAttrs
+                    NetModuleAttrs = topAttrs.netModuleAttrs
+                    AssemblyAttrs = topAttrs.assemblyAttrs
+                    DeclaredImpls = declaredImpls
+                }
+
+                // will need to pass results further somewhere
+                let _typecheckingDataResources = EncodeTypecheckingData(
+                    tcConfig,
+                    tcGlobals,
+                    tcState.Ccu,
+                    outfile,
+                    false,
+                    tcInfo)
+
+                ()
+
+            tcState, topAttrs, declaredImpls, tcEnvAtEndOfLastFile
```

Comment:
So out of 4 things, `tcState` and `tcEnvAtEndOfLastFile` will be cumbersome (yet likely possible) for pickling/unpickling - it's a lot of information. Wonder if this can be somehow worked around?